lib/modem_jwt/modem_jwt.c

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <string.h>
+#include <zephyr.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <modem/at_cmd.h>
+#include <modem/modem_jwt.h>
+
+#define JWT_CMD_TEMPLATE "AT%%JWT=%d,%u,\"%s\",\"%s\",%d,%d\r\n"
+#define MAX_INT_PRINT_SZ 11
+#define JWT_CMD_PRINT_INT_SZ (MAX_INT_PRINT_SZ * 4)
+
+/* Minimum JWT response is ~420 bytes.
+ * With very long (128 bytes) values for sub and aud claims
+ * response is ~820.
+ */
+BUILD_ASSERT(CONFIG_MODEM_JWT_MAX_LEN <= CONFIG_AT_CMD_RESPONSE_MAX_LEN,
+	     "Max JWT length exceeds max AT cmd response length");
+
+static int parse_jwt_at_cmd_resp(char *const jwt_resp)
+{
+	if (!jwt_resp) {
+		return -EINVAL;
+	}
+
+	char *jwt_start = NULL;
+	char *jwt_end = NULL;
+	size_t jwt_len;
+
+	/* Response format is:
+	 *	%JWT: "<base64_JWT>"\r\n
+	 */
+
+	/* A quotation mark indicates the start of the JWT string */
+	jwt_start = strchr(jwt_resp, '"');
+	if (!jwt_start) {
+		return -EBADMSG;
+	}
+
+	/* Move beyond first quotation mark and find second */
+	jwt_end = strchr(++jwt_start, '"');
+	if (!jwt_end) {
+		return -EBADMSG;
+	}
+
+	*jwt_end = 0;
+	jwt_len = strlen(jwt_start) + 1;
+
+	/* Move the now null-terminated jwt to the front of the buffer */
+	memmove(jwt_resp, jwt_start, jwt_len);
+
+	return 0;
+}
+
+int modem_jwt_generate(struct jwt_data *const jwt)
+{
+	if (!jwt) {
+		return -EINVAL;
+	}
+
+	int ret = 0;
+	enum at_cmd_state state;
+	char *cmd;
+	char *cmd_resp = NULL;
+	size_t cmd_sz = sizeof(JWT_CMD_TEMPLATE) + JWT_CMD_PRINT_INT_SZ +
+			(jwt->subject ? strlen(jwt->subject) : 0) +
+			(jwt->audience ? strlen(jwt->audience) : 0);
+
+	jwt->jwt_out = NULL;
+
+	ret = at_cmd_init();
+	if (ret) {
+		return -EPIPE;
+	}
+
+	/* Allocate and format the JWT request cmd */
+	cmd = k_calloc(cmd_sz, 1);
+	if (!cmd) {
+		return -ENOMEM;
+	}
+
+	ret = snprintf(cmd, cmd_sz, JWT_CMD_TEMPLATE,
+		       jwt->alg,
+		       jwt->exp_delta_s,
+		       jwt->subject ? jwt->subject : "",
+		       jwt->audience ? jwt->audience : "",
+		       jwt->sec_tag,
+		       jwt->key);
+
+	if (ret < 0 || ret >= cmd_sz) {
```

Comment:
I follow these rules in this case:
1. Do not rely on C's operator precedence rules, as they may not be obvious to those who maintain the code. To aid clarity, use parentheses to ensure proper execution order within a sequence of operations.
2. Unless it is a single identifier or constant, each operand of the logical && and || operators shall be surrounded by parentheses.

I didn't find this just now in the Zephyr or NCS coding standards when I looked for it, but in the older Nordic standard.  I thought it was an official requirement though.

Reply:
are you desiring this format:
`if ((ret < 0) || (ret >= cmd_sz))`

i prefer this way, but doesn't seem to be the standard here.

Reply:
I prefer it that way, and have been doing it that way at Nordic.  I did some spot checks in sdk-nrf and sdk-zephyr, and it is inconsistently used.